File: fortran/pr003.f90
! Problem 3
!
! 02 November 2001
!
! The prime factors of 13195 are 5, 7, 13 and 29.
!
! What is the largest prime factor of the number 600851475143 ?
! 6857

! No reason to get complex with primes, yet.

program pr003

  implicit none

  integer, parameter :: long = selected_int_kind(12)

  integer (kind=long) :: base, step

  base = 600851475143_long
  step = 2

  outer : do
    inner : do while (mod (base, step) == 0)
      base = base / step
    end do inner

    if (base == 1) exit outer

    if (step == 2) then
      step = 3
    else
      step = step + 2
    end if
  end do outer

  print *, step

end program pr003

File: rust/pr003.rs
// Problem 3
//
// 02 November 2001
//
// The prime factors of 13195 are 5, 7, 13 and 29.
//
// What is the largest prime factor of the number 600851475143 ?
//
// 6857

extern mod extra;
use sieve::Sieve;
mod sieve;

fn main() {
    let mut sieve = Sieve::new();

    let mut number = 600851475143u64;
    let mut prime = 2;

    while number > 1 {
        let p = prime as u64;
        if number % p == 0 {
            number /= p;
        } else {
            prime = sieve.next_prime(prime);
        }
    }

    println(fmt!("%u", prime));
}
